File: src/1338-Reduce-Array-Size-to-The-Half/1338.js
var minSetSize = function(arr) {
    let n = Math.floor(arr.length / 2), total = 0, res = 0;
    let freq = new Map(), cnt = new Map();
    for (let i of arr) freq.set(i, (freq.get(i) || 0) + 1);
    for (let v of freq.values()) cnt.set(v, (cnt.get(v) || 0) + 1);
    
    let cnt_keys = [...cnt.keys()];
    cnt_keys.sort((a, b) => b - a);
    for (let i of cnt_keys) {
        let dif = n - total;
        let x = Math.min(cnt.get(i), Math.ceil(dif/i));
        res += x;
        total += x * i;
        if (total >= n) return res;
    }
    return res;
};

File: src/1338-Reduce-Array-Size-to-The-Half/1338.go
func minSetSize(arr []int) int {
    n, total, res := len(arr)/2, 0, 0
    freq := make(map[int]int)
    cnt := make(map[int]int)
    for _, v := range arr {
        freq[v]++
    }
    for _, v := range freq {
        cnt[v]++
    }
    
    cnt_keys := []int{}
    for k := range cnt {
        cnt_keys = append(cnt_keys, k)
    }
    sort.Slice(cnt_keys, func(i, j int) bool {
        return cnt_keys[i] > cnt_keys[j]
    })
    
    for _, i := range cnt_keys {
        dif := n - total
        x := min(cnt[i], (dif + i - 1)/i)
        res += x
        total += x * i
        if total >= n {
            return res
        }
    }
    return res
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}